src/marqo/core/distributed_lock/zookeeper_distributed_lock.py

```diff
@@ -0,0 +1,79 @@
+from kazoo.exceptions import LockTimeout, ConnectionClosedError
+from kazoo.handlers.threading import KazooTimeoutError
+from kazoo.protocol.states import KazooState
+
+from marqo.core.distributed_lock.abstract_distributed_lock import AbstractDistributedLock
+from marqo.vespa.marqo_zookeeper_client import MarqoZookeeperClient
+from marqo.core.exceptions import BackendCommunicationError, ZooKeeperLockNotAcquiredError
+from marqo.logging import get_logger
+
+logger = get_logger(__name__)
+
+
+class ZookeeperDistributedLock(AbstractDistributedLock):
+    """A concrete implementation of distributed lock using Zookeeper."""
+
+    def __init__(self, zookeeper_client: MarqoZookeeperClient,
+                 path: str,
+                 acquire_timeout: float = 1,
+                 ):
+        """
+        Initialize the deployment lock.
+
+        Args:
+            zookeeper_client: The Zookeeper client.
+            path: The lock path.
+            acquire_timeout: The timeout to acquire the lock.
+        """
+        self._zookeeper_client = zookeeper_client
+        self._path = path
+        self._lock = self._zookeeper_client.Lock(self._path)
+        self._acquire_timeout = acquire_timeout
+
+    def acquire(self) -> bool:
+        """
+        Acquire the lock. Connect to the Zookeeper server if not connected.
+
+        Returns:
+            bool: True if the lock is acquired, raise an exception otherwise.
+        Raises:
+            BackendCommunicationError: If the Zookeeper client cannot connect to the server.
+            ZooKeeperLockNotAcquiredError: If the lock cannot be acquired within the timeout period.
+        """
+        if self._zookeeper_client.state != KazooState.CONNECTED:
+            try:
+                self._zookeeper_client.start()
+            except KazooTimeoutError as e:
+                raise BackendCommunicationError("Marqo cannot connect to backend concurrent manager. "
+                                                "Your request cannot be processed at this time. "
+                                                "Please check your network settings and try again later.") from e
+        try:
+            acquired = self._lock.acquire(timeout=self._acquire_timeout)
+            if not acquired:
+                raise ZooKeeperLockNotAcquiredError("Failed to acquire the lock.")
```

Comment:
Why/when does this happen?

Reply:
It happens when the same lock is trying to acquire twice. It will not raise a LockTimeout error but just return a False